machineconfiguration/v1alpha1/types_machineosconfig.go

```diff
@@ -0,0 +1,203 @@
+//
+// Compatibility level 4: No compatibility is provided, the API can change at any point for any reason. These capabilities should not be used by applications needing long term support.
+// +openshift:compatibility-gen:level=4
+type MachineOSConfigList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+
+	Items []MachineOSConfig `json:"items"`
+}
+
+// MachineOSConfigSpec describes user-configurable options as well as information about a build process.
+type MachineOSConfigSpec struct {
+	// machineConfigPool is the pool which the build is for
+	// +kubebuilder:validation:Required
+	MachineConfigPool MachineConfigPoolReference `json:"machineConfigPool"`
+	// buildInputs is where user options for the build live
+	// +kubebuilder:validation:Required
+	BuildInputs BuildInputs `json:"buildInputs"`
+	// +kubebuilder:validation:Required
+	// rebuild allows the user to manually trigger a rebuild
+	Rebuild MachineOSRebuild `json:"rebuild"`
+}
+
+// MachineOSConfigStatus describes the status of builds associated with this MachineOSConfig
+type MachineOSConfigStatus struct {
+	// buildHistory contains previous iterations of failed or interrupted builds related to this one.
+	// +patchMergeKey=name
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=name
+	// +kubebuilder:validation:MaxItems=2
+	// +optional
+	BuildHistory []BuildHistory `json:"buildHistory,omitempty" patchStrategy:"merge" patchMergeKey:"name"`
+}
+
+// BuildInputs holds all of the information needed to trigger a build
+type BuildInputs struct {
+	// imageBuilderType specifies the backend to be used to build the image.
+	// +kubebuilder:default:=Default
+	// +kubebuilder:validation:Enum:=Default;OpenShiftImageBuilder;PodImageBuilder
+	// Valid options are: OpenShiftImageBuilder, PodImageBuilder, and Default (OpenShiftImageBuilder)
+	ImageBuilderType MachineOSImageBuilderType `json:"imageBuilderType"`
+	// baseOSExtensionsImageURL is the base Extensions image used in the build process
+	// +kubebuilder:validation:Pattern=`^https://`
+	// +optional
+	BaseOSExtensionsImageURL string `json:"baseOSExtensionsImageURL"`
+	// baseOSImageURL is the base OSImage we use to build our custom image.
+	// +optional
+	BaseOSImageURL string `json:"baseOSImageURL"`
+	// baseImagePullSecret is the secret used to pull the base image.
+	// +kubebuilder:validation:Required
+	BaseImagePullSecret ImageSecretObjectReference `json:"baseImagePullSecret"`
+	// finalImagePushSecret is the secret used to connect to a user registry.
+	// +kubebuilder:validation:Required
+	FinalImagePushSecret ImageSecretObjectReference `json:"finalImagePushSecret"`
+	// finalImagePullSecret is the secret used to pull the final produced image.
+	// +optional
+	FinalImagePullSecret ImageSecretObjectReference `json:"finalImagePullSecret"`
+	// containerFile describes the custom data the user has specified to build into the image.
+	// +listType=set
```

Comment:
map -> key == arch, value == data

arch would be optional. both for regular clusters and multiarch

Reply:
beef up godoc